src/JuMPContainer.jl

```diff
@@ -52,68 +68,79 @@ Base.isempty(d::JuMPContainer) = isempty(_innercontainer(d))
 # 0:K -- range with compile-time starting index
 # S -- general iterable set
+export @gendict
 macro gendict(instancename,T,idxpairs,idxsets...)
     N = length(idxsets)
     allranges = all(s -> (isexpr(s,:(:)) && length(s.args) == 2), idxsets)
     if allranges
-        # JuMPArray
-        typename = symbol(string("JuMPArray",gensym()))
-        offset = Array(Int,N)
-        dictnames = Array(Symbol,N)
-        for i in 1:N
-            if isa(idxsets[i].args[1],Int)
-                offset[i] = 1 - idxsets[i].args[1]
-            else
-                error("Currently only ranges with integer compile-time starting values are allowed as index sets. $(idxsets[i].args[1]) is not an integer in range $(idxsets[i]).")
-            end
-        end
         truearray = all(idxsets) do rng
             isexpr(rng,:(:)) && # redundant, but might as well
             rng.args[1] == 1 && # starts from 1
-            length(rng.args) == 2 || rng.args[2] == 1 # steps of one
+            (length(rng.args) == 2 || rng.args[2] == 1) # steps of one
         end
-        typecode = :(type $(typename){T} <: JuMPArray{T,$N,$truearray}; innerArray::Array{T,$N}; name::String;
-                            indexsets; indexexprs::Vector{IndexPair} end)
-        getidxlhs = :(Base.getindex(d::$(typename)))
-        setidxlhs = :(Base.setindex!(d::$(typename),val))
-        getidxrhs = :(Base.getindex(d.innerArray))
-        setidxrhs = :(Base.setindex!(d.innerArray,val))
-        maplhs = :(Base.map(f::Function,d::$(typename)))
-        if truearray # return a julia Array here
-            maprhs = :(map(f,d.innerArray))
+        if truearray
+            sizes = Expr(:tuple, [esc(rng.args[end]) for rng in idxsets]...)
+            :($(esc(instancename)) = Array($T, $sizes))
         else
-            maprhs = :($(typename)(map(f,d.innerArray),d.name,d.indexsets,d.indexexprs))
-        end
-        wraplhs = :(JuMPContainer_from(d::$(typename),inner)) # helper function that wraps array into JuMPArray of similar type
-        wraprhs = :($(typename)(inner, d.name, d.indexsets, d.indexexprs))
-        for i in 1:N
-            varname = symbol(string("x",i))
+            typename = symbol(string("JuMPArray",gensym()))
+            dictnames = Array(Symbol,N)
+            # JuMPArray
+            offset = Array(Int,N)
+            for i in 1:N
+                if isa(idxsets[i].args[1],Int)
+                    offset[i] = 1 - idxsets[i].args[1]
+                else
+                    error("Currently only ranges with integer compile-time starting values are allowed as index sets. $(idxsets[i].args[1]) is not an integer in range $(idxsets[i]).")
+                end
+            end
+            typecode = quote
+                type $(typename){T} <: JuMPArray{T,$N}
+                    innerArray::Array{T,$N}
+                    meta::Dict{Symbol,Any}
+                end
+            end
+            constrlhs = :($(typename)(innerArray::Array))
```

Comment:
Could maybe make this (and a lot of the other methods here?) a single method for ``JuMPArray``, outside the ``eval`` block.